llama_index/indices/query/base.py

```diff
@@ -12,15 +13,22 @@
 
 
 class BaseQueryEngine(ABC):
+    def __init__(self, callback_manager: CallbackManager) -> None:
+        self.callback_manager = callback_manager or CallbackManager([])
+
     def query(self, str_or_query_bundle: QueryType) -> RESPONSE_TYPE:
-        if isinstance(str_or_query_bundle, str):
-            str_or_query_bundle = QueryBundle(str_or_query_bundle)
-        return self._query(str_or_query_bundle)
+        with self.callback_manager.as_trace("query"):
```

Comment:
curious: what happens when we nest `as_trace`?

For example, if I use `as_trace` context manager to wrap retrieve calls as well? I might wan to do this because e.g. I only use retrieve, without the response synthesis.

Reply:
nested as_traces are ignored and the events under the nested traces are added to the current ongoing run. 

This allows many entry points to a "run" (from_documents, raw index constructor, insert, query, etc.) and retains clear boundaries on a run

Reply:
gotcha, this is great.